x-pack/plugins/observability_solution/asset_manager/server/lib/entities/install_entity_definition.ts

```diff
@@ -0,0 +1,157 @@
+    ingestPipelines: {
+      history: false,
+      latest: false,
+    },
+    transforms: {
+      history: false,
+      latest: false,
+    },
+    definition: false,
+  };
+
+  try {
+    logger.debug(`Installing definition ${JSON.stringify(definition)}`);
+    const entityDefinition = await saveEntityDefinition(soClient, definition);
+    installState.definition = true;
+
+    // install ingest pipelines
+    logger.debug(`Installing ingest pipelines for definition ${definition.id}`);
+    await createAndInstallHistoryIngestPipeline(esClient, entityDefinition, logger, spaceId);
+    installState.ingestPipelines.history = true;
+    await createAndInstallLatestIngestPipeline(esClient, entityDefinition, logger, spaceId);
+    installState.ingestPipelines.latest = true;
+
+    // install transforms
+    logger.debug(`Installing transforms for definition ${definition.id}`);
+    await createAndInstallHistoryTransform(esClient, entityDefinition, logger);
+    installState.transforms.history = true;
+    await createAndInstallLatestTransform(esClient, entityDefinition, logger);
+    installState.transforms.latest = true;
+
+    return entityDefinition;
+  } catch (e) {
+    logger.error(`Failed to install entity definition ${definition.id}`, e);
+    // Clean up anything that was successful.
+    if (installState.definition) {
+      await deleteEntityDefinition(soClient, definition, logger);
+    }
+
+    if (installState.ingestPipelines.history) {
+      await deleteHistoryIngestPipeline(esClient, definition, logger);
+    }
+    if (installState.ingestPipelines.latest) {
+      await deleteLatestIngestPipeline(esClient, definition, logger);
+    }
+
+    if (installState.transforms.history) {
+      await stopAndDeleteHistoryTransform(esClient, definition, logger);
+    }
+
+    if (installState.transforms.latest) {
+      await stopAndDeleteLatestTransform(esClient, definition, logger);
+    }
+
+    throw e;
+  }
+}
+
+export async function installBuiltInEntityDefinitions({
+  esClient,
+  soClient,
+  logger,
+  builtInDefinitions,
+  spaceId,
+}: Omit<InstallDefinitionParams, 'definition'> & { builtInDefinitions: EntityDefinition[] }) {
+  if (builtInDefinitions.length === 0) return;
+
+  logger.debug(`Starting installation of ${builtInDefinitions.length} built-in definitions`);
+  const installPromises = builtInDefinitions.map(async (builtInDefinition) => {
+    const definitions = await findEntityDefinitions({
+      esClient,
+      soClient,
+      id: builtInDefinition.id,
+    });
+
+    if (definitions.length === 0) {
+      return await installAndStartDefinition({
+        definition: builtInDefinition,
+        esClient,
+        soClient,
+        logger,
+        spaceId,
+      });
+    }
+
+    const definition = definitions[0];
+    // verify current installation
+    if (!definition.state.installed) {
+      logger.debug(`Detected partial installation of definition [${definition.id}], reinstalling`);
```

Comment:
this state is - the definition is 'saved', but maybe either the transforms or ingest pipelines are not there?

is this state possible to get into? i would prefer if we just make it impossible to be in this state than have defensive code like this.

Reply:
This state can be when
- there is a failure to cleanup the components when there we fail to install a definition https://github.com/elastic/kibana/blob/main/x-pack/plugins/observability_solution/asset_manager/server/routes/entities/create.ts#L89
- user manually deletes a transform/ingest pipeline